utils/lib/rr_graph/graph.py

```diff
@@ -1508,168 +1814,246 @@ def add_nodes_for_pin_class(self, block, pin_class, switch, sides):
-                self.add_edge(
-                    int(src_node.get("id")), int(pin_node.get("id")),
-                    int(switch.get("id")))
+                self.create_rr_edge_with_nodes(src_node, pin_node, switch)
 
         else:
             assert False, "Unknown dir of {} for {}".format(
                 pin_class.direction, str(pin_class))
 
-        #return pin_node
+    def create_nodes_from_block(self, block, switch, sides):
+        """
+        Creates the SOURCE/SINK rr_nodes for each pin class
+        Creates the IPIN/OPIN rr_nodes for each pin inside a pin class.
+        Creates the rr_edges which connect these together.
+
+        Parameters
+        ----------
+        block : Block
+        switch : Switch
+        sides : RoutingNodeSide
+
+        Examples
+        --------
+        """
+        for pc in block.block_type.pin_classes:
+            self.create_nodes_from_pin_class(block, pc, switch, sides)
 
-    def add_nodes_for_block(self, block, switch, sides):
+    def create_nodes_from_blocks(self, blocks, switch, sides):
         """
-        Creates the SOURCE/SINK nodes for each pin class
-        Creates the IPIN/OPIN nodes for each pin inside a pin class.
+        Parameters
+        ----------
+        block : Block
+        switch : Switch
+        sides : RoutingNodeSide
+        """
+        for block in blocks:
+            self.create_nodes_from_block(block, switch, sides)
 
-        >>> test_add_nodes_for_block()
+    def create_node_from_track(self, track):
         """
-        for pc in block.block_type.pin_classes:
-            self.add_nodes_for_pin_class(block, pc, switch, sides)
+        Creates the CHANX/CHANY rr_node for a Track object.
 
-    def add_node_for_track(self, track):
+        Parameters
+        ----------
+        track : channels.Track
+
+        Examples
+        --------
+        """
         assert_type(track, Track)
         assert track.idx != None
 
-        return self.add_node(
+        return self.create_rr_node(
             track.start,
             track.end,
             track.idx,
             track.type.value,
             direction=track.direction,
-            segment_id=track.segment.id)
+            segment_id=track.segment_id)
+
+
+
+def sides_always_right(*a, **kw):
+    return RoutingNodeSide.RIGHT
 
 
 class Graph:
-    '''
+    """
     Top level representation, holds the XML root
     For <rr_graph> node
-    '''
-
-    def __init__(self, rr_graph_file=None, verbose=False,
-             clear_fabric=False):
-        '''
-        clear_fabric: remove channels, all nodes, and all edges
-        FIXME: need pin nodes to get placement, so not clearing any nodes for now
-        Long term we need to reshuffle these instead of deleting and re-adding
-        '''
+    """
+
+    def __init__(
+            self,
+            rr_graph_file=None,
+            verbose=False,
+            clear_fabric=False,
+            sides=sides_always_right):
```

Comment:
Thoughts on pushing back on VPR devs about side placement? I feel this is something we shouldn't have to track

Reply:
I think we should log a bug about this.